## Entendimiento del problema
### By: Maricel Karina Martínez Hernández

### Date: 2026-08-15

### Description:

Este notebook contiene el entendimiento del problema y la obtención inicial de los datos para el proyecto de predicción de precios de viviendas en Boston.

### Issue:

#### Requerimiento
Crear una nueva rama de git (Usar Gitflow) y Crear un notebook para la obtención de datos tipo RAW.

crear la rama a partir de un issue: https://docs.github.com/es/issues/tracking-your-work-with-issues/using-issues/creating-a-branch-for-an-issue

Realizar los pasos de: https://joserzapata.github.io/post/ciencia-datos-proyecto-python/1-data/

Descargar los datos y entender el problema a realizar, contestar en el notebook

¿Cual es el objetivo del problema?
¿Cómo se usará su solución?
¿Cuáles son las soluciones actuales (si las hay)?
¿Cómo se debe enmarcar este problema (supervisado / no supervisado, en línea / fuera de línea, etc.)
¿Cómo se debe medir el desempeño o el rendimiento de la solución, una primera intuicion?
¿La medida de desempeño está alineada con el objetivo del problema?
¿Cuál sería el desempeño o rendimiento mínimo necesario para alcanzar el objetivo del problema?
¿Cuáles son los problemas parecidos? ¿Se puede reutilizar experiencias o herramientas ya creadas?
¿Hay experiencia del problema disponible?
(Importante) ¿Cómo se puede resolver el problema manualmente?
Hacer un listado de los supuestos que hay hasta este momento.
Cual es la fuente de los datos?
Como se actualizan los datos?
Cada cuanto tiempo se actualizan los datos
#### Entregables
Notebook para obtención de los datos.

Se debe realizar un Pull request para ingresar el notebook a la ramamain para esto debe tener mínimo 1 revisiones de otras personas del Curso y que pase los check del CI/CD


## 📚 Importar  librerías

In [4]:
# base libraries for data science
from pathlib import Path

import pandas as pd

## 💾 Cargar datos

In [7]:
data_path = Path.cwd().resolve().parents[1] / "data" / "01_raw" / "Precios_Casas_Boston.csv"

boston_df = pd.read_csv(data_path, low_memory=False)
boston_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       448 non-null    int64  
 1   crim     443 non-null    float64
 2   zn       446 non-null    float64
 3   indus    442 non-null    float64
 4   chas     446 non-null    float64
 5   nox      444 non-null    float64
 6   rm       442 non-null    float64
 7   age      444 non-null    float64
 8   dis      442 non-null    float64
 9   rad      444 non-null    float64
 10  tax      443 non-null    float64
 11  ptratio  446 non-null    float64
 12  black    445 non-null    float64
 13  lstat    447 non-null    float64
 14  medv     432 non-null    float64
dtypes: float64(14), int64(1)
memory usage: 52.6 KB


## 🔍Exploración inicial de los datos

Se realiza una exploración inicial del conjunto de datos para conocer su
estructura, las variables disponibles y algunas de las observaciones
contenidas en el archivo.

In [8]:
boston_df.sample(10)

,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
66,97,0.11504,0.0,2.89,0.0,445.0,6163.00,69.6,3.4952,2.0,276.0,18.0,391.83,11.34,21.4
90,134,0.32982,0.0,NaN,0.0,624.0,5822.00,95.4,2.4699,4.0,437.0,21.2,388.69,15.03,18.4
18,31,1.13081,0.0,8.14,0.0,538.0,5713.00,94.1,4233.0000,4.0,307.0,21.0,360.17,22.60,12.7
422,155,1.41385,0.0,19.58,1.0,871.0,6129.00,96.0,1.7494,5.0,403.0,14.7,321.02,15.12,17.0
439,485,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,20.2,370.73,13.34,NaN
252,384,7.99248,0.0,18.10,0.0,0.7,5.52,100.0,1.5331,24.0,666.0,20.2,396.90,24.56,12.3
408,134,0.32982,0.0,21.89,0.0,624.0,5822.00,95.4,2.4699,4.0,437.0,21.2,388.69,15.03,18.4
60,88,0.07151,0.0,4.49,0.0,449.0,6121.00,56.8,3.7476,3.0,247.0,18.5,395.15,8.44,22.2
147,214,0.14052,0.0,10.59,0.0,489.0,6375.00,32.3,3.9454,4.0,277.0,18.6,385.81,9.38,28.1
300,462,3.69311,0.0,18.10,0.0,713.0,6376.00,88.4,2.5671,24.0,666.0,20.2,391.43,14.65,17.7


## 💡Información de los datos

El conjunto de datos contiene información relacionada con características
socioeconómicas, ambientales y estructurales de diferentes zonas de Boston.
La variable objetivo del proyecto es `medv`, que representa el valor medio
de las viviendas.

Antes de continuar con el procesamiento, se revisan las variables disponibles
y sus tipos de datos.

In [9]:
boston_df.columns

Index(['ID', 'crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad',
       'tax', 'ptratio', 'black', 'lstat', 'medv'],
      dtype='str')

## ✅Identificación de variables no predictivas

La variable `ID` corresponde a un identificador de los registros y no representa
una característica relacionada con las viviendas. Por esta razón, no se
utilizará como variable predictora.

Mantener un identificador como predictor podría introducir relaciones
artificiales en el modelo y afectar su capacidad de generalización.

In [10]:
columns_to_use = [
    "crim",
    "zn",
    "indus",
    "chas",
    "nox",
    "rm",
    "age",
    "dis",
    "rad",
    "tax",
    "ptratio",
    "black",
    "lstat",
    "medv",
]

## 🧠Selección de variables

Para la primera versión del conjunto de datos se excluye la variable `ID`,
debido a que funciona únicamente como identificador de los registros y no
aporta información predictiva sobre el precio de las viviendas.

La variable `medv` se conserva porque corresponde al objetivo que se desea
predecir.

Las demás variables se mantienen como variables predictoras candidatas para
las siguientes etapas del proyecto.

In [11]:
boston_final_df = boston_df[columns_to_use]

boston_final_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     443 non-null    float64
 1   zn       446 non-null    float64
 2   indus    442 non-null    float64
 3   chas     446 non-null    float64
 4   nox      444 non-null    float64
 5   rm       442 non-null    float64
 6   age      444 non-null    float64
 7   dis      442 non-null    float64
 8   rad      444 non-null    float64
 9   tax      443 non-null    float64
 10  ptratio  446 non-null    float64
 11  black    445 non-null    float64
 12  lstat    447 non-null    float64
 13  medv     432 non-null    float64
dtypes: float64(14)
memory usage: 49.1 KB


## 🔎Entendimiento del problema

### ¿Cuál es el objetivo del problema?

El objetivo del proyecto es desarrollar un modelo de aprendizaje automático que permita predecir el valor medio de las viviendas (medv) a partir de diferentes características socioeconómicas, ambientales y relacionadas con las viviendas y su entorno en diferentes zonas de Boston.

La variable medv corresponde a la variable objetivo que se desea estimar, mientras que las demás variables, excepto ID, se consideran inicialmente como variables predictoras candidatas.

### ¿Cómo se usará su solución?

La solución se utilizará para realizar estimaciones del valor medio de las viviendas a partir de las características disponibles de cada registro.

El modelo podrá servir como una herramienta de apoyo para estimar precios y analizar la relación entre las características socioeconómicas, ambientales y estructurales de una zona y el valor de las viviendas.

La predicción obtenida por el modelo debe entenderse como una estimación basada en los datos disponibles y no como una valoración inmobiliaria definitiva.

### ¿Cuáles son las soluciones actuales?

Actualmente, la estimación del valor de una vivienda puede realizarse mediante métodos tradicionales de valoración inmobiliaria y mediante modelos estadísticos y de aprendizaje automático.

Para este proyecto se plantea utilizar aprendizaje automático con el fin de aprender automáticamente la relación entre las características de las zonas y el valor medio de las viviendas.

### ¿Cómo se debe enmarcar este problema?

El problema se plantea como un problema de aprendizaje supervisado, debido a que se dispone de una variable objetivo conocida (medv) que se desea predecir.

Es específicamente un problema de regresión, ya que la variable objetivo es numérica.

Inicialmente se considera un problema fuera de línea (offline), debido a que se trabajará con un conjunto de datos histórico y estático. El modelo será entrenado utilizando los datos disponibles y posteriormente será evaluado sobre datos que no hayan sido utilizados durante el entrenamiento.

### ¿Cómo medir el desempeño?

Al tratarse de un problema de regresión, se utilizarán métricas adecuadas para comparar los valores reales de medv con los valores predichos por el modelo.

Como primera aproximación se consideran:

MAE (Mean Absolute Error): mide el error absoluto promedio de las predicciones.
RMSE (Root Mean Squared Error): mide la magnitud de los errores y da mayor peso a errores grandes.
R²: indica qué proporción de la variabilidad de la variable objetivo puede ser explicada por el modelo.

Como métrica principal se propone inicialmente utilizar MAE, complementada con RMSE y R².

### ¿La medida está alineada con el objetivo?

Sí. La métrica propuesta está alineada con el objetivo porque el propósito del modelo es obtener estimaciones de medv lo más cercanas posible a sus valores reales.

El MAE permite medir directamente la magnitud promedio del error de las predicciones y se expresa en las mismas unidades de la variable objetivo. El RMSE complementa esta evaluación al dar mayor importancia a los errores grandes, mientras que R² permite conocer la capacidad explicativa del modelo.

### ¿Cuál es el rendimiento mínimo necesario?

En esta etapa no es posible establecer un valor numérico definitivo para el desempeño mínimo requerido, debido a que todavía no se ha construido ni evaluado un modelo.

Como punto de referencia se establecerá primero un modelo base (baseline), que permita obtener un desempeño inicial. Los modelos posteriores deberán mejorar este resultado.

El desempeño mínimo aceptable se determinará después de comparar los diferentes modelos y analizar el error obtenido en relación con el objetivo de predicción.

### ¿Cuáles son los problemas parecidos?

La predicción de precios o valores de viviendas es un problema clásico de regresión. Existen múltiples experiencias previas utilizando técnicas estadísticas y algoritmos de aprendizaje automático para estimar precios a partir de características de las viviendas y de su entorno.

Por esta razón, es posible reutilizar experiencias, metodologías y herramientas desarrolladas para problemas similares, especialmente algoritmos de regresión, técnicas de preparación de datos y métricas de evaluación.

### ¿Hay experiencia del problema disponible?

Sí. La predicción del valor de viviendas es un problema ampliamente estudiado y existen conjuntos de datos, metodologías y modelos previamente desarrollados para este tipo de problema.

Esta experiencia puede utilizarse como referencia para seleccionar técnicas de regresión, estrategias de preparación de datos y métricas de evaluación.

### ¿Cómo se puede resolver manualmente?

Una forma manual de resolver el problema consistiría en comparar una vivienda o zona con otras que presenten características similares y utilizar los valores observados de dichas viviendas como referencia para realizar una estimación.

Para ello podrían analizarse características como el número promedio de habitaciones (rm), la proporción de población de menor condición socioeconómica (lstat), la tasa de criminalidad (crim), la antigüedad de las viviendas (age), la distancia a determinados centros (dis) y otras variables disponibles.

El problema de realizar este procedimiento manualmente es que resulta difícil considerar simultáneamente todas las variables y sus relaciones. El aprendizaje automático busca automatizar este proceso identificando patrones en los datos y utilizándolos para generar nuevas predicciones.

### Supuestos

1. medv representa la variable objetivo que se desea predecir.
2. Las variables distintas de ID y medv contienen información potencialmente útil para realizar la predicción.
3. ID funciona únicamente como identificador y no contiene información predictiva relevante.
4. Los registros disponibles son representativos del problema que se desea modelar.
5. Las relaciones observadas entre las variables predictoras y medv pueden utilizarse para realizar predicciones.
6. Los datos requieren una etapa posterior de preparación antes del entrenamiento de los modelos.
7. Los valores faltantes y las posibles inconsistencias identificadas en la exploración deben ser tratados antes de construir el modelo.
8. El conjunto de datos puede dividirse posteriormente en datos de entrenamiento y evaluación para medir la capacidad de generalización del modelo.
9. Las variables predictoras disponibles son suficientes para obtener una estimación útil de medv.

### ¿Cuál es la fuente de los datos?

Los datos utilizados en este proyecto fueron proporcionados por el docente en formato CSV y corresponden a una versión modificada del conjunto de datos Boston Housing.

El archivo utilizado como fuente RAW es Precios_Casas_Boston.csv, almacenado en la carpeta data/01_raw/.

En esta etapa se conserva la información tal como fue proporcionada, sin realizar procesos de limpieza o transformación.

### ¿Cómo se actualizan los datos?

Los datos utilizados en este proyecto corresponden a un conjunto de datos estático proporcionado por el docente. Por lo tanto, no existe un mecanismo automático de actualización mediante una API, base de datos en tiempo real u otra fuente dinámica.

Cualquier modificación o actualización del conjunto de datos dependería de una nueva versión proporcionada por el docente.

### ¿Cada cuánto se actualizan?

No existe una frecuencia de actualización periódica establecida para este conjunto de datos. Para efectos del proyecto, se considera una fuente de datos estática.


## 📊 Analisis de resultados y conclusiones


La exploración inicial permitió identificar la estructura y las características principales del conjunto de datos utilizado para el proyecto. La base contiene 448 registros y 15 variables, de las cuales ID corresponde a un identificador y medv corresponde a la variable objetivo. Las demás variables se consideran inicialmente como posibles variables predictoras.

Durante la revisión inicial se identificaron valores faltantes en diferentes variables. En particular, medv presenta 432 valores no nulos, por lo que existen registros sin información para la variable objetivo. También se observan valores faltantes en varias de las variables predictoras.

La exploración de algunas observaciones también permite identificar valores que deberán ser revisados durante la etapa de preparación de los datos. Por ejemplo, se observan valores como 538 en nox, 6575 en rm y 4233 en dis, así como registros con múltiples valores faltantes.

Estos hallazgos no se corrigen en esta etapa, debido a que el objetivo del notebook es realizar la obtención y el entendimiento inicial de los datos RAW. Sin embargo, deberán ser considerados en la siguiente etapa del proyecto para determinar las estrategias apropiadas de limpieza y preparación.

También se determinó que la variable ID no debe utilizarse como variable predictora, debido a que corresponde únicamente a un identificador de los registros. Por esta razón, se excluyó del conjunto de variables que serán utilizadas en las siguientes etapas.

A partir de esta exploración, el problema queda definido como un problema de aprendizaje supervisado de regresión, cuyo objetivo será predecir medv a partir de las características disponibles.

Como siguiente paso se propone realizar la preparación de los datos, incluyendo el análisis detallado de los valores faltantes, la revisión de posibles inconsistencias, el tratamiento de los registros que lo requieran y la posterior división de los datos para entrenamiento y evaluación del modelo.

## 💡 Propuestas e ideas



A partir de los resultados obtenidos se proponen las siguientes actividades para continuar el proyecto:

Analizar con mayor detalle los valores faltantes y definir una estrategia adecuada para su tratamiento.
Revisar los valores de las variables para identificar posibles inconsistencias o valores anómalos.
Analizar la distribución de la variable objetivo medv.
Estudiar la relación entre las variables predictoras y medv.
Dividir los datos en conjuntos de entrenamiento y evaluación evitando utilizar información del conjunto de evaluación durante el entrenamiento.
Construir un modelo base de regresión que permita establecer una referencia inicial de desempeño.
Comparar diferentes algoritmos de regresión utilizando MAE, RMSE y R².
Analizar los errores obtenidos para determinar las fortalezas y limitaciones de los modelos.

Estas actividades permitirán avanzar desde el entendimiento inicial de los datos hacia las etapas de preparación, modelado y evaluación del proyecto.